File: app/src/main/java/com/onedelay/mymovie/database/MovieDao.java
package com.onedelay.mymovie.database;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

@Dao
public interface MovieDao {
    @Insert
    void insertMovies(MovieEntity... movieEntities);

    @Update
    void updateMovies(MovieEntity... movieEntities);

    @Query("SELECT * FROM Movies")
    List<MovieEntity> selectDataMovies();

    @Query("SELECT * FROM Movies")
    LiveData<List<MovieEntity>> selectMovies();

    /* Room 에서는 SQL 인젝션를 방지하기 위해 ORDER BY 를 파라미터로 넘길 수 없다.
     * 참고 : https://stackoverflow.com/questions/44240906/android-room-order-by-not-working */
    @Query("SELECT * FROM Movies ORDER BY reservation_rate DESC")
    LiveData<List<MovieEntity>> selectMoviesOrderByReservationRate();

    @Query("SELECT * FROM Movies ORDER BY audience_rating DESC")
    LiveData<List<MovieEntity>> selectMoviesOrderByAudienceRating();

    @Query("SELECT * FROM Movies ORDER BY date DESC")
    LiveData<List<MovieEntity>> selectMoviesOrderByDate();

    @Query("SELECT * FROM Movies WHERE id = :id")
    LiveData<MovieEntity> selectMovieDetailLive(int id);

    @Query("SELECT * FROM Movies WHERE id = :id")
    MovieEntity selectMovieDetail(int id);

    @Query("DELETE FROM Movies")
    void clear();
}

File: app/src/main/java/com/onedelay/mymovie/data/local/dao/MovieDao.kt
package com.onedelay.mymovie.data.local.dao

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Insert
import android.arch.persistence.room.Query
import android.arch.persistence.room.Update
import com.onedelay.mymovie.data.local.entity.MovieEntity

@Dao
interface MovieDao {
    @Insert
    fun insertMovies(vararg movieEntities: MovieEntity)

    @Update
    fun updateMovies(vararg movieEntities: MovieEntity)

    @Query("SELECT * FROM Movies")
    fun selectDataMovies(): List<MovieEntity>

    @Query("SELECT * FROM Movies")
    fun selectMovies(): LiveData<List<MovieEntity>>

    /* Room 에서는 SQL 인젝션를 방지하기 위해 ORDER BY 를 파라미터로 넘길 수 없다.
     * 참고 : https://stackoverflow.com/questions/44240906/android-room-order-by-not-working */
    @Query("SELECT * FROM Movies ORDER BY reservation_rate DESC")
    fun selectMoviesOrderByReservationRate(): LiveData<List<MovieEntity>>

    @Query("SELECT * FROM Movies ORDER BY audience_rating DESC")
    fun selectMoviesOrderByAudienceRating(): LiveData<List<MovieEntity>>

    @Query("SELECT * FROM Movies ORDER BY date DESC")
    fun selectMoviesOrderByDate(): LiveData<List<MovieEntity>>

    @Query("SELECT * FROM Movies WHERE id = :id")
    fun selectMovieDetailLive(id: Int): LiveData<MovieEntity>

    @Query("SELECT * FROM Movies WHERE id = :id")
    fun selectMovieDetail(id: Int): MovieEntity

    @Query("DELETE FROM Movies")
    fun clear()
}
